mozilla-conduit/lando-ui
# Add Asset Pipeline with Flask Assets

### Add asset pipeline using flask-assets library

Lando UI was missing a way to allow developers to write css and javascript files and have those files all be compiled to one file which would be served to the user on page load. Without this functionality we would have to include every single css / javascript file by hand in the html - which would be a strain on productivity.

Flask-Assets [https://flask-assets.readthedocs.io/en/latest/](https://flask-assets.readthedocs.io/en/latest/) is a library which helps us solve this problem. It allows us to organize our code in different files, write a manifest file defining how those files should be compiled into one big file, and allows us to do some pre/post processing on the files.

This commit organizes our css and javascript code underneath the standard flask static folder. Inside is a src folder which contains the js and css files as well as an 'assets.yml' file. The assets.yml file is where we define how everything should be combined together. Flask-Assets provides a Jinja template helper that will automatically generate a link to the bundled asset when you want to include it in the html.

This commit also initialize our asset pipeline with some defaults and some examples. Notably Semantic-UI is included (along with jQuery which it depends on). The home page has also been revamped to demonstrate how to use the new asset pipeline.

### Add support for building production static assets

In development, the development flask server serves static assets and further it will recompile assets when assets have changed between requests.

In production we don't want to do this. Ideally we want to compile the static assets once and serve them always. This commit solves the two problems needed to achieve this:

1) This integrates flask-assets cli tools which exposes a command `flask assets build` which builds and outputs the assets. This command is run in the development image so that the production image can be free of build dependencies.

2) It updates the production image and circle.yml to copy in the static assets. These assets will remain unchanged until the next production image is built. This means that the assets could be served via proxy or volume mounted and served via nginx.

This also fixes the problem with the circle yml not pushing to dockerhub correctly 😸 

# Steps for testing
1. Clone and checkout this branch.
2. You will see a file structure like this 
![landoui_file_structure](https://cloud.githubusercontent.com/assets/7828780/25852436/52658c8e-3497-11e7-9be1-3c653c597379.PNG)
3. You can add new routes to the web application in the `pages.py` file, you can modify existing templates in the `templates` folder, and you can add CSS, Javascript, and LESS CSS in the `static/src` folder. There are examples of this in the project.
4. When modifying assets, you will have to update the `static/src/assets.yml` file if there is anything more you want to do to the file other than include in in the bundled assets.
5. See `templates/layout.html` for an example of how to include a link to a bundled asset.
6. For initial testing: In the root of the project, run `docker-compose up`. By default this starts a server at your dockermachine's IP address at port `7777`. You should see something like 
![landoui-intial](https://cloud.githubusercontent.com/assets/7828780/25852598/e242f63e-3497-11e7-9497-abf95bb44bcb.PNG)
7. Inspect the source HTML to confirm that the assets links are properly being created as expected, also inspect the source files themselves to ensure they are being bundled as expected (expected = js minified, and LESS CSS compiled to normal css).
8. Try adding new javascript and css assets to ensure they work as expected. 
9. Since [Semantic UI](https://semantic-ui.com/introduction/getting-started.html) is bundled by default, try using some of the elements on the home page to make sure they work properly.

To test that assets can be prebuilt for productions run the build command exactly like the `circle.yml` runs it: `docker-compose run -e FLASK_APP=/app/landoui/assets_app.py lando-ui flask assets build`. (Because of the volume mount in the docker-compose file) This should create a folder called build in your `landoui/static/build` folder. Ensure that these assets are bundled as expected.
